server/filestore.go

```diff
@@ -1586,54 +1593,42 @@ func (fs *fileStore) recoverFullState() (rerr error) {
 		return errCorruptState
 	}
 
-	// Grab the max blk index we see from scanning the directory. The full snapshot has the index that was lmb when
-	// we created it, so with that and max we know blocks to process. We do this in parallel in casee lots of blks.
-	blkFiles := <-rch
-
-	defer func() {
-		// Make sure we saw all of our blk files.
-		for _, mb := range fs.blks {
-			if _, ok := blkFiles.blksSeen[mb.index]; !ok {
-				if ld, _, _ := mb.rebuildState(); ld != nil {
-					// If we have lost data make sure we track here.
-					fs.addLostData(ld)
-					rerr = errCorruptState
-				}
-			}
-		}
-	}()
-
 	// Move into place our state, msgBlks and subject info.
 	fs.state = state
 
-	// If our saved state is past what we see on disk, fallback and rebuild.
-	if blkFiles != nil && blkFiles.maxIndex < blkIndex {
-		return errPriorState
-	}
-
 	// First let's check the happy path, open the blk file that was the lmb when we created the full state.
 	// See if we have the last block available.
 	var matched bool
 	var mb *msgBlock
 	if mb = fs.bim[blkIndex]; mb != nil {
-		matched = bytes.Equal(mb.lastChecksum(), lchk[:])
-		if matched && blkIndex == blkFiles.maxIndex {
-			return nil
+		if _, err := os.Stat(mb.mfn); err != nil && os.IsNotExist(err) {
+			// If our saved state is past what we see on disk, fallback and rebuild.
+			if ld, _, _ := mb.rebuildState(); ld != nil {
+				fs.addLostData(ld)
+			}
+			return errPriorState
+		}
+
+		if matched = bytes.Equal(mb.lastChecksum(), lchk[:]); !matched {
+			// Remove the last message block since we will re-process below.
+			fs.removeMsgBlockFromList(mb)
 		}
-		// Remove the last message block since we will re-process below.
-		fs.removeMsgBlockFromList(mb)
 	}
 
-	// If we are here we did not match the happy path.
-	// We need to go through and find our checksum. This should be in blkIndex, but might not be.
-	start, stop := blkIndex, blkFiles.maxIndex
+	// We may need to check other blocks. Even if we matched last checksum we will see if there is another block.
+	// If we did not match we re-process the last block.
+	start := blkIndex
 	if matched {
 		start++
 	}
 
-	for bi := start; bi <= stop; bi++ {
+	for bi := start; ; bi++ {
```

Comment:
Minor nit but `for bi := start; true; bi++ {` might be clearer.

Reply:
I believe that is the way Go recommends it so will leave as is.